src/test_memcachedsolution.cpp

```diff
@@ -0,0 +1,864 @@
+ */
+
+#include "gtest/gtest.h"
+
+#include "memcachedstore.h"
+#include "processinstance.h"
+
+#include <vector>
+#include <iostream>
+#include <fstream>
+#include <stdio.h>
+
+static const SAS::TrailId DUMMY_TRAIL_ID = 0x12345678;
+static const int BASE_MEMCACHED_PORT = 33333;
+static const int ASTAIRE_PORT = 11311;
+
+// Fixture for all memcached solution tests.
+//
+// This fixture:
+// - Creates a unique key and an instance of TopologyNeutralMemcachedStore for
+//   each test.
+// - Tidies up instances of memcached and Astaire and any config files between
+//   sets of tests.
+// - Provides helper methods for memcached operations and managing instances of
+//   memcached and Astaire
+class BaseMemcachedSolutionTest : public ::testing::Test
+{
+public:
+  /// Kill and clear all the memcached and Astaire instances, and remove the
+  /// cluster_settings file.
+  static void TearDownTestCase()
+  {
+    for (std::vector<std::shared_ptr<MemcachedInstance>>::iterator inst = _memcached_instances.begin();
+         inst != _memcached_instances.end();
+         ++inst)
+    {
+      (*inst)->kill_instance();
+    }
+    _memcached_instances.clear();
+
+    for (std::vector<std::shared_ptr<AstaireInstance>>::iterator inst = _astaire_instances.begin();
+         inst != _astaire_instances.end();
+         ++inst)
+    {
+      (*inst)->kill_instance();
+    }
+    _astaire_instances.clear();
+
+    if (remove("cluster_settings") != 0)
+    {
+      perror("remove cluster_settings");
+    }
+  }
+
+  /// Create a new store and a new unique key for this test.
+  virtual void SetUp()
+  {
+    _store = new TopologyNeutralMemcachedStore();
+
+    // Create a new key for every test (to prevent tests from interacting with
+    // each other).
+    _key = std::to_string(_next_key++);
+  }
+
+  /// Make sure all the memcached and Astaire instances are running before
+  /// starting finishing the test.
+  virtual void TearDown()
+  {
+    EXPECT_TRUE(wait_for_instances());
+
+    delete _store; _store = NULL;
+  }
+
+  /// Helper method for generating a new unique key in the middle of a test.
+  void get_new_key()
+  {
+    _key = std::to_string(_next_key++);
+  }
+
+  /// Creates and starts up the specified number of memcached instances. Also
+  /// sets up the appropriate cluster_settings file (which just contains a
+  /// single line of the form:
+  ///
+  ///   servers=127.0.0.1:33333,127.0.0.1:33334,...
+  ///
+  static void create_and_start_memcached_instances(int memcached_instances)
+  {
+    std::ofstream cluster_settings("cluster_settings");
+
+    for (int ii = 0; ii < memcached_instances; ++ii)
+    {
+      if (ii == 0)
+      {
+        cluster_settings << "servers=";
+      }
+      else
+      {
+        cluster_settings << ",";
+      }
+
+      // Each instance should listen on a new port.
+      int port = BASE_MEMCACHED_PORT + ii;
+      _memcached_instances.emplace_back(new MemcachedInstance(port));
```

Comment:
Is `emplace_back` really necessary here? Would `puch_back` work? I ask because I had to look up `emplace_back` so there is some cognitive overhead to using it.